cmd/query/app/grpc_handler_test.go

```diff
@@ -134,335 +148,318 @@ func newGRPCServer(t *testing.T, q *querysvc.QueryService, logger *zap.Logger, t
+
+	q := querysvc.NewQueryService(spanReader, dependencyReader,
+		querysvc.QueryServiceOptions{
+			ArchiveSpanReader: archiveSpanReader,
+			ArchiveSpanWriter: archiveSpanWriter,
+		})
+
 	logger := zap.NewNop()
 	tracer := opentracing.NoopTracer{}
 
-	grpcServer, addr := newGRPCServer(t, q, logger, tracer)
+	server, addr := newGRPCServer(t, q, logger, tracer)
 
-	return grpcServer, addr, readStorage, dependencyStorage, archiveSpanReader, archiveSpanWriter
+	return &grpcServer{
+		server:            server,
+		lisAddr:           addr,
+		spanReader:        spanReader,
+		depReader:         dependencyReader,
+		archiveSpanReader: archiveSpanReader,
+		archiveSpanWriter: archiveSpanWriter,
+	}
+}
+
+func withServerAndClient(t *testing.T, actualTest func(server *grpcServer, client *grpcClient)) {
+	server := initializeTestServerGRPCWithOptions(t)
+	client := newGRPCClient(t, server.lisAddr)
+	defer server.server.Stop()
+	defer client.conn.Close()
+
+	actualTest(server, client)
 }
 
 func TestGetTraceSuccessGRPC(t *testing.T) {
-	server, addr, readMock, _ := initializeTestServerGRPC(t)
-	defer server.Stop()
-	readMock.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).
-		Return(mockTrace, nil).Once()
+	withServerAndClient(t, func(server *grpcServer, client *grpcClient) {
 
-	client, conn := newGRPCClient(t, addr)
-	defer conn.Close()
+		server.spanReader.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).
+			Return(mockTrace, nil).Once()
 
-	res, err := client.GetTrace(context.Background(), &api_v2.GetTraceRequest{
-		TraceID: mockTraceIDgrpc,
-	})
-	spanResChunk, _ := res.Recv()
+		res, err := client.GetTrace(context.Background(), &api_v2.GetTraceRequest{
+			TraceID: mockTraceIDgrpc,
+		})
+
+		spanResChunk, _ := res.Recv()
 
-	assert.NoError(t, err)
-	assert.Equal(t, spanResChunk.Spans[0].TraceID, mockTraceID)
+		assert.NoError(t, err)
+		assert.Equal(t, spanResChunk.Spans[0].TraceID, mockTraceID)
+
+	})
 }
 
 func TestGetTraceDBFailureGRPC(t *testing.T) {
-	server, addr, readMock, _ := initializeTestServerGRPC(t)
-	defer server.Stop()
-	readMock.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).
-		Return(nil, errStorageGRPC).Once()
+	withServerAndClient(t, func(server *grpcServer, client *grpcClient) {
 
-	client, conn := newGRPCClient(t, addr)
-	defer conn.Close()
+		server.spanReader.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).
+			Return(nil, errStorageGRPC).Once()
 
-	res, err := client.GetTrace(context.Background(), &api_v2.GetTraceRequest{
-		TraceID: mockTraceIDgrpc,
-	})
-	assert.NoError(t, err)
+		res, err := client.GetTrace(context.Background(), &api_v2.GetTraceRequest{
+			TraceID: mockTraceIDgrpc,
+		})
+		assert.NoError(t, err)
+
+		spanResChunk, err := res.Recv()
 
-	spanResChunk, err := res.Recv()
+		assert.Error(t, err)
```

Comment:
Great work!

This was on my TODO but I couldn't get around to fix it, but we should probably improve the error checking. Instead of just checking for the presence of an error like so - 

```
assert.Error(t, err)
```
we should also check for the right error message - 
```
import google.golang.org/grpc/status
.
.
expectedError := status.Errorf(2, "Storage error")
assert.Equal(t, expectedErr, err)
```

Would you be able to do it in this PR?